// @fortawesome/angular-fontawesome/esm5/testing/testing.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { FaIconLibrary, FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { MockFaIconLibrary } from './icon/mock-icon-library.service';
import * as ɵngcc0 from '@angular/core';
var FontAwesomeTestingModule = /** @class */ (function () {
    function FontAwesomeTestingModule() {
    }
FontAwesomeTestingModule.ɵfac = function FontAwesomeTestingModule_Factory(t) { return new (t || FontAwesomeTestingModule)(); };
FontAwesomeTestingModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: FontAwesomeTestingModule });
FontAwesomeTestingModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [{ provide: FaIconLibrary, useExisting: MockFaIconLibrary }], imports: [FontAwesomeModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FontAwesomeTestingModule, [{
        type: NgModule,
        args: [{
                exports: [FontAwesomeModule],
                providers: [{ provide: FaIconLibrary, useExisting: MockFaIconLibrary }]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FontAwesomeTestingModule, { exports: function () { return [FontAwesomeModule]; } }); })();
    return FontAwesomeTestingModule;
}());
export { FontAwesomeTestingModule };
