// @fortawesome/angular-fontawesome/fesm5/fortawesome-angular-fontawesome-testing.js
import { __decorate } from 'tslib';
import { ɵɵdefineInjectable, Injectable, NgModule } from '@angular/core';
import { FontAwesomeModule, FaIconLibrary } from '@fortawesome/angular-fontawesome';

import * as ɵngcc0 from '@angular/core';
var dummyIcon = {
    prefix: 'fad',
    iconName: 'dummy',
    icon: [512, 512, [], 'f030', 'M50 50 H462 V462 H50 Z'],
};
var MockFaIconLibrary = /** @class */ (function () {
    function MockFaIconLibrary() {
    }
    MockFaIconLibrary.prototype.addIcons = function () {
        throw new Error('Attempt to add an icon to the MockFaIconLibrary.');
    };
    MockFaIconLibrary.prototype.addIconPacks = function () {
        throw new Error('Attempt to add an icon pack to the MockFaIconLibrary.');
    };
    MockFaIconLibrary.prototype.getIconDefinition = function (prefix, name) {
        return dummyIcon;
    };
    MockFaIconLibrary.ɵprov = ɵɵdefineInjectable({ factory: function MockFaIconLibrary_Factory() { return new MockFaIconLibrary(); }, token: MockFaIconLibrary, providedIn: "root" });
MockFaIconLibrary.ɵfac = function MockFaIconLibrary_Factory(t) { return new (t || MockFaIconLibrary)(); };
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MockFaIconLibrary, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
    return MockFaIconLibrary;
}());

var FontAwesomeTestingModule = /** @class */ (function () {
    function FontAwesomeTestingModule() {
    }
FontAwesomeTestingModule.ɵfac = function FontAwesomeTestingModule_Factory(t) { return new (t || FontAwesomeTestingModule)(); };
FontAwesomeTestingModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: FontAwesomeTestingModule });
FontAwesomeTestingModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [{ provide: FaIconLibrary, useExisting: MockFaIconLibrary }], imports: [FontAwesomeModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FontAwesomeTestingModule, [{
        type: NgModule,
        args: [{
                exports: [FontAwesomeModule],
                providers: [{ provide: FaIconLibrary, useExisting: MockFaIconLibrary }]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FontAwesomeTestingModule, { exports: function () { return [FontAwesomeModule]; } }); })();
    return FontAwesomeTestingModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { FontAwesomeTestingModule, MockFaIconLibrary as ɵa };
